services/wallet/database.go

```diff
@@ -790,31 +789,206 @@ func insertBlocksWithTransactions(creator statementCreator, account common.Addre
 	return nil
 }
 
-func insertRange(creator statementCreator, account common.Address, network uint64, from *big.Int, to *big.Int) (err error) {
-	insert, err := creator.Prepare("INSERT INTO blocks_ranges (network_id, address, blk_from, blk_to) VALUES (?, ?, ?, ?)")
+func upsertRange(creator statementCreator, account common.Address, network uint64, from *big.Int, to *big.Int) (err error) {
+	update, err := creator.Prepare(`UPDATE blocks_ranges
+                SET blk_to = ?
+                WHERE address = ?
+                AND network_id = ?
+                AND blk_to = ?`)
+
 	if err != nil {
 		return err
 	}
 
-	_, err = insert.Exec(network, account, (*SQLBigInt)(from), (*SQLBigInt)(to))
+	res, err := update.Exec((*SQLBigInt)(to), account, network, (*SQLBigInt)(from))
+
 	if err != nil {
 		return err
 	}
+	affected, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if affected == 0 {
+		insert, err := creator.Prepare("INSERT INTO blocks_ranges (network_id, address, blk_from, blk_to) VALUES (?, ?, ?, ?)")
+		if err != nil {
+			return err
+		}
+
+		_, err = insert.Exec(network, account, (*SQLBigInt)(from), (*SQLBigInt)(to))
+		if err != nil {
+			return err
+		}
+	}
 
 	return
 }
 
+type BlocksRange struct {
+	from *big.Int
+	to   *big.Int
+}
+
+func (db *Database) getOldRanges(account common.Address, network uint64) ([]*BlocksRange, error) {
+	query := `select blk_from, blk_to from blocks_ranges
+	          where address = ?
+	          and network_id = ?
+	          order by blk_from`
+
+	rows, err := db.db.Query(query, account, db.network)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	ranges := []*BlocksRange{}
+	for rows.Next() {
+		from := &big.Int{}
+		to := &big.Int{}
+		err = rows.Scan((*SQLBigInt)(from), (*SQLBigInt)(to))
+		if err != nil {
+			return nil, err
+		}
+
+		ranges = append(ranges, &BlocksRange{
+			from: from,
+			to:   to,
+		})
+	}
+
+	return ranges, nil
+}
+
+func getNewRanges(ranges []*BlocksRange) ([]*BlocksRange, []*BlocksRange) {
+	initValue := big.NewInt(-1)
+	prevFrom := big.NewInt(-1)
+	prevTo := big.NewInt(-1)
+	hasMergedRanges := false
+	newRanges := []*BlocksRange{}
+	deletedRanges := []*BlocksRange{}
```

Comment:
Although this will compile fine, it is best to avoid an empty slice declaration via a literal, instead use a nil slice declaration.

See https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

```suggestion
	var newRanges []*BlocksRange
	var deletedRanges []*BlocksRange
```

Reply:
fixed